daemon/cmd/kube_proxy_replacement.go

```diff
@@ -34,9 +34,17 @@ import (
 	"github.com/cilium/cilium/pkg/probe"
 	"github.com/cilium/cilium/pkg/safeio"
 	"github.com/cilium/cilium/pkg/sysctl"
+	"github.com/cilium/cilium/pkg/version"
+	"github.com/cilium/cilium/pkg/versioncheck"
 	wgTypes "github.com/cilium/cilium/pkg/wireguard/types"
 )
 
+var (
+	// Host routing on the L3 devices counts on a patch for bpf_skb_change_head: https://lore.kernel.org/bpf/20210519154743.2554771-2-joamaki@gmail.com/
+	// The patch is available as of kernel 5.13-rc4, so we simply check kernel version to guarantee that.
+	hasSkbChangeHeadPatch = versioncheck.MustCompile(">=5.13.0")
```

Comment:
We don't usually rely on specific kernel versions because that breaks too easily (e.g., RHEL kernels). Instead we probe the specific kernel feature we need if possible, or something that was introduced at the same time (same patchset ideally, same release otherwise).

In this case, it's not a feature but a bugfix. I'm expecting it will be backported to all relevant LTS kernels (?). Do we even need to enable this conditionally? What happens if the kernel doesn't support it? How obvious is the failure today?

Reply:
> I'm expecting it will be backported to all relevant LTS kernels

This is correct, I just checked the latest LTS kernels, 4.14.307, 4.19.275,  5.4.234, 5.10.172, and all of them have included the patch.

> What happens if the kernel doesn't support it?

As far as I can see, the skb->data will point to the middle of IP header without the patch, so the skb will be dropped by cilium's tc bpf, typically by `validate_ethertype` in `from-container`.

Reply:
So Cilium would likely drop those packets because they are malformed. Maybe it's fine to just enable this by default without any probing? We can ask in the community meeting if unsure.

Reply:
The other option is to probe for a helper which went into the same release as Jussi's fixes and only enable it then. This would ignore the stable releases for the time being, but at least it would ensure there are no gaps in between.

Reply:
I'm proposing dropping the probe mostly because it's fairly complicated to implement (for a probe) and this feels like a fairly corner case scenario already.

Reply:
Sure. To implement a probe looks fun, I could do it in future if really necessary.

Reply:
I'm going to resolve this conversation by now. The conclusion so far is to make no check and to presume costumers are using TLS kernels which have been backported with necessary patches. If it turns out to be needed in future, I can implement a feature probe then.

Reply:
Just for documentation purposes:

- `bpf: Set mac_len in bpf_skb_change_head` landed in kernel `5.13`
- bpf-host-routing is available since kernel 5.10
- between 5.10 and 5.13, the only kernel still maintained on kernel.org is `5.10`.
- `bpf: Set mac_len in bpf_skb_change_head` was backported to `5.10.42`